test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -123,6 +123,227 @@ void testExprSimple02() {
+  Stmt* simplified = IRSimplifier::simplify(stmt);
+  Block* body = dynamic_cast<Block*>(simplified);
+  ASSERT_EQ(body->nstmts(), 2);
+  auto biter = body->begin();
+
+  For* loop = dynamic_cast<For*>(*biter++);
+  assertForRange(loop, 0, 6);
+
+  loop = dynamic_cast<For*>(*biter);
+  assertForRange(loop, 6, 10);
+}
+
+void testExprSplitAndSlice() {
+  // 0: splitWithTail
+  // 1: sliceTail on inner loop
+  // 2: sliceHead on outer loop
+  KernelScope kernel_scope;
+  auto func = [](const ExprHandle& x) {
+    return ExprHandle(1.0f) + cast<float>(x);
+  };
+  Tensor* tensor = Compute("f", {{100, "x"}}, func);
+  LoopNest l({tensor});
+
+  For* outer;
+  For* inner;
+  For* tail;
+  std::vector<For*> loops = l.getLoopStmtsFor(tensor);
+  // outer: [0, 4)
+  // inner: [0, 21)
+  // tail:  [84, 100)
+  l.splitWithTail(loops[0], 21, &outer, &inner, &tail);
+
+  For* inner_head;
+  For* inner_tail;
+  l.sliceTail(inner, 2, &inner_head, &inner_tail);
+
+  For* outer_head;
+  For* outer_tail;
+  l.sliceHead(outer, 2, &outer_head, &outer_tail);
+
+  Stmt* stmt = l.root_stmt();
+  Stmt* simplified = IRSimplifier::simplify(stmt);
+  Block* body = dynamic_cast<Block*>(simplified);
+  // for (int x_outer = 0; x_outer < 2; x_outer++) {
+  //   for (int x_inner = 0; x_inner < 19; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  //   for (int x_inner = 19; x_inner < 21; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  // }
+  // for (int x_outer = 2; x_outer < 4; x_outer++) {
+  //   for (int x_inner = 0; x_inner < 19; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  //   for (int x_inner = 19; x_inner < 21; x_inner++) {
+  //     f[21 * x_outer + x_inner] = 1.f + float(21 * x_outer + x_inner);
+  //   }
+  // }
+  // for (int x_tail = 0; x_tail < 16; x_tail++) {
+  //   f[x_tail + 84] = 1.f + float(x_tail + 84);
+  // }
+  ASSERT_EQ(body->nstmts(), 3);
+  auto biter = body->begin();
+
+  outer_head = dynamic_cast<For*>(*biter++);
+  assertForRange(outer_head, 0, 2);
+  {
+    // Inside outer_head.
+    Block* body = outer_head->body();
+    ASSERT_EQ(body->nstmts(), 2);
+    auto biter = body->begin();
+
+    For* inner_head = dynamic_cast<For*>(*biter++);
+    assertForRange(inner_head, 0, 19);
+
+    For* inner_tail = dynamic_cast<For*>(*biter);
+    assertForRange(inner_tail, 19, 21);
+  }
+
+  outer_tail = dynamic_cast<For*>(*biter++);
+  assertForRange(outer_tail, 2, 4);
+  {
+    // Inside outer_tail.
+    Block* body = outer_tail->body();
+    ASSERT_EQ(body->nstmts(), 2);
+    auto biter = body->begin();
+
+    For* inner_head = dynamic_cast<For*>(*biter++);
+    assertForRange(inner_head, 0, 19);
+
+    For* inner_tail = dynamic_cast<For*>(*biter);
+    assertForRange(inner_tail, 19, 21);
+  }
+
+  tail = dynamic_cast<For*>(*biter);
+  assertForRange(tail, 0, 16);
+}
+
```

Comment:
A few more cases worth testing:
* Slice then Normalize tail.
* SliceTail, then SliceHead on the tail. (tests case where start is not zero)
* over-Slice (ie slice by a larger factor than the loop extent).
* Slice a loop with a non constant bound (ie, the loop extent is a Var).

Reply:
Added more tests.